lib/record_store/provider/ns1.rb

```diff
@@ -0,0 +1,192 @@
+require 'ns1'
+
+module RecordStore
+  class Provider::NS1 < Provider
+    class << self
+      def client
+        Provider::NS1::Client.new(api_key: secrets['api_key'])
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout)
+        records_for_zone(zone)
+         .map{|short_record| build_from_api(short_record, zone)}
+         .flatten
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.zones.map do |zone|
+          zone["zone"]
+        end
+      end
+
+      private
+
+      # Fetches simplified records for the provided zone
+      def records_for_zone(zone)
+        client.zone(zone)["records"]
+      end
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        new_answers = [{ answer: [ build_api_answer_from_record(record) ]}]
+
+        record_fqdn = record.fqdn.sub(/.$/, '')
+
+        existing_record = client.record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type
+        )
+
+        if existing_record.nil?
+          client.create_record(
+            zone: zone,
+            fqdn: record_fqdn,
+            type: record.type,
+            params: { answers: new_answers, ttl: record.ttl }
+          )
+          return
+        end
+
+        existing_answers = existing_record["answers"].map { |answer| symbolize_keys(answer) }
+        client.modify_record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type,
+          params: { answers: existing_answers + new_answers }
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = record.fqdn.sub(/.$/, '')
+
+        existing_record = client.record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type
+        )
+        return if existing_record.nil?        
+        
+        pruned_answers = existing_record["answers"]
+          .map { |answer| symbolize_keys(answer) }
+          .reject { |answer| answer[:answer] == build_api_answer_from_record(record) } 
```

Comment:
This seems a bit weak if we consider that a record might have multiple entries of the same `answer` (with different `meta` for instance). But I guess this is not something to be considered in the `record_store` case as we're only handling static records with very simple setups and no situation where we would need or have `meta` fields. Is that right ?

Reply:
That's right. But I can definitely test with some complicated setups that you can think of.

Reply:
My idea would be that if there is a chance we need to handle the meta thing, then yes, let's take time for this. But if it's not something that can happen with the record store in the current state, we can ignore this for now (as anyway, in this case, supporting that case would require more work elsewhere, and the addition to this provider could be done at the same time).

So my question to you would be: is this a case that can happen (having multiple identical answers with different metas) with the current state of the record store ?

Reply:
I think for now, we should assume that record store will the exclusive maintainer of the zones it's touching.  If we don't allow it to create complicated annotated records, it shouldn't be expected to maintain them.

If we unleash record store on a zone that has already been populated, ... the behaviour should be undefined.  To tighten this up, I'd go so far as blacklisting fields on answers.  If an answer has a `meta` field, `raise` because record_store must not have created it.